### PR TITLE
Controller

### PR DESCRIPTION
**Pull request by EzraWeller**
adding comments

### CODE REVIEW

File: contracts/ContinuousTokenFactory.sol [hunk 1]
```diff
@@ -2,9 +2,164 @@
...
+    return true;
+  }
+
+  event Buy(
+    bytes32 id, 
+    string name,
+    uint256 amount,
+    uint256 price,
+    address buyer,
+    address owner
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: I get why we need this, but I'm thinking the names should be `buyer->from` and `owner->to` if we need both?

File: contracts/ContinuousTokenFactory.sol [hunk 2]
```diff
@@ -2,9 +2,164 @@
...
+  );
+
+  function _buy(
+    bytes32 id,
+    uint256 amount,
+    uint256 maxPrice,
+    address buyer,
+    address owner
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: I get why we need this, but I'm thinking the names should be `buyer->from` and `owner->to` if we need both?

File: contracts/ContinuousTokenFactory.sol [hunk 3]
```diff
@@ -2,9 +2,164 @@
...
+    uint256 amount,
+    uint256 price,
+    address buyer,
+    address owner
+  );
+
+  function _buy(
+    bytes32 id,
+    uint256 amount,
+    uint256 maxPrice,
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: So, here's a thought: now that these are internal, do we need this max price pattern in the internal buy function at all? Why not have `_buy` just take `(id, price, from, to)` and then mint whatever amount that buys?

One reason to do this is that we don't really care what the amount is when buying license, so we'd no longer have to bother with it as an input.

File: contracts/SquadController.sol [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+import "./ContinuousTokenFactory.sol";
+import "./Curve.sol";
+import "@nomiclabs/buidler/console.sol";
+
+contract SquadController is Ownable, ContinuousTokenFactory {
+    using SafeMath for uint256;
+    using SafeMath for uint16;
+
+    TokenClaimCheck public tokenClaimCheck;
+
```
Comment (jessebmiller) [inline contracts/SquadController.sol:37]: Add allow list of curves rather than take it in from the client

File: contracts/SquadController.sol [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    uint16 public networkFee; // in basis points
+    uint16 public maxNetworkFee; // in basis points
+
+    address public treasury;
+    mapping(address => uint256) public acounts;
```
Comment (jessebmiller) [inline contracts/SquadController.sol]: accounts (typo)

File: contracts/SquadController.sol [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                totalPrice >= purchasePrice,
+                "SquadController: not enough to meet purchasePrice"
+                );
+
+        // buy `amount` of the continuous token to be claimed by this license
+        _buy(id, amount, totalPrice, msg.sender, address(this));
+
+        // Create a license to claim those tokens check for the caller
+        continuousTokens[id].token.approve(address(tokenClaimCheck), amount);
+        uint256 license = tokenClaimCheck.mint(
```
Comment (jessebmiller) [inline contracts/SquadController.sol]: let's call this `licenseId` rather than license

File: contracts/SquadController.sol [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+                             msg.sender,
+                             amount,
+                             address(this),
+                             address(token),
+                             tokenURI
+                             );
+
+        // record valid license
+        validLicenses[license] = id;
```
Comment (jessebmiller) [inline contracts/SquadController.sol]: Let's name this id `contributionId`

File: contracts/ContinuousTokenFactory.sol [hunk 8]
```diff
@@ -2,9 +2,168 @@
...
+        address buyer,
+        address owner
+    );
+
+    function buy(
+        bytes32 id,
+        uint256 amount,
+        address buyer,
+        address owner
+    ) external onlyOwner mustExist(id) returns (bool) {
```
Comment (jessebmiller) [inline contracts/ContinuousTokenFactory.sol]: Use consistent ordering of onlyOwner and mustExist

File: contracts/ContinuousTokenFactory.sol [hunk 9]
```diff
@@ -2,9 +2,168 @@
...
+        uint256 amount,
+        uint256 price,
+        address buyer,
+        address owner
+    );
+
+    function buy(
+        bytes32 id,
+        uint256 amount,
+        address buyer,
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: buyer --> from
owner --> to

File: contracts/ContinuousTokenFactory.sol [hunk 10]
```diff
@@ -2,9 +2,168 @@
...
+        string name,
+        uint256 amount,
+        uint256 price,
+        address seller
+    );
+
+    function sell(
+        bytes32 id,
+        uint256 amount,
+        address seller
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: implement `from` and `to` here as well

File: contracts/ContinuousTokenFactory.sol [hunk 11]
```diff
@@ -2,9 +2,168 @@
...
+        return reserveToken.transfer(to, amount);
+    }
+
+    function reserveBalanceOf(address account) public view returns (uint256) {
+        return reserveToken.balanceOf(account);
+    }
+
+    // TODO Consider changing this to priceOf to avoid confusion with
+    // price variables in function bodies
+    function price(
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: change name to `priceOf`

File: contracts/ContinuousTokenFactory.sol [hunk 12]
```diff
@@ -2,9 +2,168 @@
...
+        bytes32 id,
+        uint256 supply,
+        uint256 units
+    ) public view mustExist(id) returns (uint256) {
+        Curve curve = continuousTokens[id].curve;
+        return curve.price(supply, units);
+    }
+
+    function totalSupply(bytes32 id) public view returns (uint256) {
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: `totalSupplyOf`

File: contracts/ContinuousTokenFactory.sol [hunk 13]
```diff
@@ -2,9 +2,168 @@
...
+        return continuousTokens[id].token.totalSupply();
+    }
+
+    function tokenAddress(bytes32 id)
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: `tokenAddressOf`

File: contracts/SquadController.sol [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+        uint256 amount,
+        uint256 price
+    );
+
+    function buyLicense(
+        bytes32 contributionId,
+        uint256 amount,
+        uint256 maxPrice
+    ) external mustExist(contributionId) {
+        ERC20 token = ERC20(tokenFactory.tokenAddress(contributionId));
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: tokenFactory is storing an ERC20--can we have it return that directly?

File: contracts/SquadController.sol [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+        uint256 supply = token.totalSupply();
+        uint256 totalPrice = price(contributionId, supply, amount);
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: `price = priceOf`

File: contracts/SquadController.sol [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+        }
+        return account == tokenClaimCheck.ownerOf(licenseId);
+    }
+
+    function sellContinuousTokens(
+        bytes32 contributionId,
+        uint256 amount,
+        uint256 minPrice
+    ) public mustExist(contributionId) {
+        uint256 supply = tokenFactory.totalSupply(contributionId);
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: `totalSupplyOf`

File: contracts/SquadController.sol [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+            "SquadController: nothing to withdraw"
+        );
+        FeeLib.FeeSplit memory feeSplit = FeeLib.calculateFeeSplit(
+            networkFeeRate,
+            accounting.total(account)
+        );
+        uint256 withdrawAmount = feeSplit.remainder;
+
+        // transfer to account address
+        tokenFactory.transferReserve(account, withdrawAmount);
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: can just directly reference `feeSplit.remainder`

File: contracts/SquadController.sol [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+
+        // transfer to treasury
+        tokenFactory.transferReserve(treasury, feeSplit.fee);
+
+        emit Withdraw(account, withdrawAmount, feeSplit.fee);
+    }
+
+    function reserveDust() public view returns (uint256) {
+        return
+            tokenFactory.reserveBalanceOf(address(this)).sub(
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: this dust would be in the tokenFactory, not the controller? Investigate!

File: contracts/SquadController.sol [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+                accounting.accountsTotal()
+            );
+    }
+
+    function recoverReserveDust() public {
+        tokenFactory.transferReserve(treasury, reserveDust());
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: add a require that `reserveDust()` > 0

File: contracts/SquadController.sol [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    function price(
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: `priceOf`

File: contracts/SquadController.sol [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+        bytes32 contributionId,
+        uint256 supply,
+        uint256 amount
+    ) public view returns (uint256) {
```
Comment (EzraWeller) [inline contracts/SquadController.sol:448]: see if we can make this external

File: contracts/SquadController.sol (hunk 14 above)
Comment (EzraWeller) [reply contracts/SquadController.sol]: We can use the controller's `tokenAddress` function and possibly remove that function from `continuousTokenFactory` entirely.

File: contracts/ContinuousTokenFactory.sol [hunk 22]
```diff
@@ -2,9 +2,168 @@
 
 pragma solidity >=0.6.0 <0.7.0;
 
+import "@openzeppelin/contracts/math/SafeMath.sol";
```
Comment (EzraWeller) [inline contracts/ContinuousTokenFactory.sol]: We can use `BondingCurveFactory`!

File: contracts/SquadController.sol [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+import "@nomiclabs/buidler/console.sol";
+
+contract SquadController is Ownable {
+    using SafeMath for uint256;
+    using SafeMath for uint16;
+
+    TokenClaimCheck public tokenClaimCheck;
+    Curve public curve;
+    Accounting public accounting;
+    ContinuousTokenFactory public tokenFactory;
```
Comment (EzraWeller) [inline contracts/SquadController.sol]: Let's call this `bcFactory` to reflect the new name.

**Pull request merged**